src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -3726,7 +3726,7 @@ public void visitReference(final JCMemberReference that) {
 
             if (!env.info.attributionMode.isSpeculative && that.getMode() == JCMemberReference.ReferenceMode.NEW) {
                 Type enclosingType = exprType.getEnclosingType();
-                if (enclosingType != null && enclosingType.hasTag(CLASS)) {
+                if (enclosingType != null && enclosingType.hasTag(CLASS) && exprType.tsym.hasOuterInstance()) {
```

Comment:
Looks good - in fact using `getEnclosingType` has a single test is almost always wrong...

Reply:
That said... I'm not too sure we should do that here... e.g. if I have some Foo::new, and Foo is a class with an enclosing instance, then we should check that the enclosing instance is indeed available from the context where the method reference is used... otherwise that could lead to a situation where there's a missing check?

Reply:
This is a good example of ambiguous terminology in the compiler's internal naming of things. What does "implicit this" mean here?

Clearly the term "implicit this" includes any _immediate_ outer instance, but does it include a 2nd outer instance, which is what we are dealing with here? 

The empirical evidence seems to imply no. I say that because, _without_ this change, you get this bogus error:
```
$ cat CtorRefError.java
class CtorRefError {
    class Inner {
        Inner() {
            class Local { }
            Runnable r = Local::new;
            super();
        }
    }
}
$ javac --enable-preview --release 24 CtorRefError.java 
CtorRefError.java:5: error: cannot reference this before supertype constructor has been called
            Runnable r = Local::new;
                         ^
```
(That error is incorrect because per the new JLS local classes in early construction contexts are supposed to behave just like anonymous classes in that they have no _immediate_ enclosing instance.)

So it looks like the "implicit this" that `resolveImplicitThis()` is looking for (and not finding) is always the immediate outer instance (`Inner`), not the 2nd outer instance - because the 2nd outer instance `CtorRefError` is there and available.

_With_ this change, you get the NPE described in [JDK-8333313](https://bugs.openjdk.org/browse/JDK-8333313) as expected. So this part of the patch is really a fix for a related bug, not the actual bug described in the Jira ticket.

Reply:
FYI, I've moved this particular fix to a separate bug/PR: https://github.com/openjdk/jdk/pull/19705